### PR TITLE
Coercion

### PR DESCRIPTION
**Pull request by sarakolding**
- [x] I have battle-tested on Overtaci (RMAPPS1279)

## Notes for reviewers
Added medication loaders for opioid dependence medications.
Also added functionality for subsetting by administration route and/or method of administration. These are added as args in the loaders, and can be passed for groups of features when the loader_kwargs issue in timeseriesflattener is fixed.
Please disregard the naming of the branch and the first commits (sorry about that).

### CODE REVIEW

File: src/psycop_feature_generation/loaders/raw/utils.py [hunk 1]
```diff
@@ -131,6 +135,12 @@ def load_from_codes(
         + f" FROM [fct].{fct} WHERE {source_timestamp_col_name} IS NOT NULL AND ({match_col_sql_str})"
     )
 
+    if administration_method:
+        sql += f" AND type_kodetekst = '{administration_method}'"
+
+    if administration_route:
+        sql += f" AND admvej_kodetekst = '{administration_route}'"
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/loaders/raw/utils.py:142]: Looks great! Would be super helpful if we guard against invalid `administration_methods`, so users are warned if they set requirements that will return no rows.

```suggestion
    if administration_method:
        allowed_administration_methods = ("OR", "IM", "IV")
        if administration_method not in allowed_administration_methods:
            raise ValueError(f"Administration method is not allowed, allowed values are {allowed_administration_methods}")
        sql += f" AND type_kodetekst = '{administration_method}'"
```

Similar for `administration_route`.

Does that make sense? :-)

**Review by MartinBernstorff — CHANGES_REQUESTED** [review]
Looks great! Tiny comments.

**Comment by MartinBernstorff** [comment]
Another question: How does this work if we request prescribed medication as well? Since they don't have an `administration_method`, we'll return no rows. Probably want to either handle that or raise an error if both `load_prescribed = true` and an `administration_method` is set.

File: src/psycop_feature_generation/loaders/raw/utils.py (hunk 1 above)
Comment (sarakolding) [reply src/psycop_feature_generation/loaders/raw/utils.py:142]: Yes, agreed. great idea!

**Comment by sarakolding** [comment]
> Another question: How does this work if we request prescribed medication as well? Since they don't have an `administration_method`, we'll return no rows. Probably want to either handle that or raise an error if both `load_prescribed = true` and an `administration_method` is set.

Hadn't thought of that at all. Thanks! I'll make the necessary changes :-)

File: src/psycop_feature_generation/loaders/raw/utils.py [hunk 2]
```diff
@@ -78,6 +78,8 @@ def load_from_codes(
     match_with_wildcard: bool = True,
     n_rows: Optional[int] = None,
     exclude_codes: Optional[list[str]] = None,
+    administration_route: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (MartinBernstorff) [reply src/psycop_feature_generation/loaders/raw/utils.py:81]: Ah, that sucks! An alternate route is to log a warning if the function returns 0 rows. Probably useful in general 👍

File: src/psycop_feature_generation/loaders/raw/load_medications.py [hunk 3]
```diff
@@ -38,11 +40,20 @@ def load(
...
         pd.DataFrame: Cols: dw_ek_borger, timestamp, {atc_code_prefix}_value = 1
     """
 
+    if load_prescribed == True and (
+        administration_method is not None or administration_route is not None
+    ):
+        raise TypeError(
+            "load() got an unexpected combination of arguments. When load_prescribed=True, administration_method and administration_route must be NoneType objects."
+        )
+
```
Comment (sarakolding) [inline src/psycop_feature_generation/loaders/raw/load_medications.py:56]: I'm not completely sure that this is the correct error type.

File: src/psycop_feature_generation/loaders/raw/load_medications.py [hunk 4]
```diff
@@ -140,25 +155,39 @@ def concat_medications(
 
 # data_loaders primarly used in psychiatry
 @data_loaders.register("antipsychotics")
-def antipsychotics(n_rows: Optional[int] = None) -> pd.DataFrame:
+def antipsychotics(
+    n_rows: Optional[int] = None,
+    load_prescribed: Optional[bool] = True,
+    load_administered: Optional[bool] = True,
```
Comment (sarakolding) [inline src/psycop_feature_generation/loaders/raw/load_medications.py:161]: added load_prescribed and _administered as args, but kept the default the same as it was previously.

File: src/psycop_feature_generation/loaders/raw/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarakolding) [reply src/psycop_feature_generation/loaders/raw/utils.py:85]: I sort of compromised, so you get a warning that it returns 0 rows, but it also tells you what the allowed options are. If it is bad practice, I can change it to just a general warning when returning 0 rows, and then people can google for themselves what the options could be ;-)

File: src/psycop_feature_generation/loaders/raw/load_medications.py (hunk 3 above)
Comment (MartinBernstorff) [reply src/psycop_feature_generation/loaders/raw/load_medications.py:56]: Fair question, I'm not sure either. `TypeError` seems fine to me.

File: src/psycop_feature_generation/loaders/raw/load_medications.py [hunk 5]
```diff
@@ -40,13 +40,20 @@ def load(
...
+        administration_route (str, optional): Whether to subset by a specific administration route, e.g. 'OR', 'IM' or 'IV'. Only applicable for administered medication, not prescribed. Defaults to None.
+        administration_method (str, optional): Whether to subset by method of administration, e.g. 'PN' or 'Fast'. Only applicable for administered medication, not prescribed. Defaults to None.
 
     Returns:
         pd.DataFrame: Cols: dw_ek_borger, timestamp, {atc_code_prefix}_value = 1
     """
 
+    if load_prescribed == True and (
+        administration_method is not None or administration_route is not None
+    ):
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/loaders/raw/load_medications.py:52]:
```suggestion
    if load_prescribed and any([administration_method, administration_route]):
```

Tiny tiny thing, just makes it more pythonic and brief :-)

**Review by MartinBernstorff — APPROVED** [review]
Looks great! Thanks for doing this 👍

**Comment by MartinBernstorff** [comment]
Excellent! Feel free to merge when you've resolved the merge conflicts. They're conflicts with your own commits, so I expect you're more than qualified 😉

**Pull request merged**